### ISSUE TITLE
twisted.web.guard not ported to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7974]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| edorum2 reported|
|-|-|
|Trac ID|trac#7974|
|Type|enhancement|
|Created|2015-07-20 12:01:06Z|
|Branch|https://github.com/twisted/twisted/tree/py3-t.web.guard-7974-2|

The twisted.web.guard and it's implementation in twisted.web._auth has not been ported to Python 3. 

Attachments:

* [web-auth-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/40f/40fdbfb5de31b4973be0bc08945c613341bc1d80/fe0114c536121d3a1acb96e1721085639b0dd21f.patch) (263 bytes) - added by edorum2 on 2015-07-20 12:11:34Z - 
* [web-auth-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/40f/40fdbfb5de31b4973be0bc08945c613341bc1d80/94f49fdfe4b19d2877a6ab377a54c4e4f274fe17.patch) (22470 bytes) - added by edorum2 on 2015-07-20 12:29:00Z - 
* [web-auth-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/40f/40fdbfb5de31b4973be0bc08945c613341bc1d80/8a6171db2f5ee4bf252b48796afdf966deb8a2e5.patch) (22463 bytes) - added by edorum2 on 2015-07-23 10:47:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7974 7974
type__enhancement enhancement
reporter__edorum2 edorum2
priority__normal normal
milestone__ 
branch__branches_py3_t_web_guard_7974_2 branches/py3-t.web.guard-7974-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1437393666632087 1437393666632087
changetime__1437666099504695 1437666099504695
version__None None
owner__edorum2 edorum2

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|edorum2 removed owner|
|-|-|

I added twisted.web.guard to the modules list in dist3.py, even if there are no tests for it. But then it only does imports from twisted.web_auth where there are tests. 
I was a bit unsure about if it should have been in the almostModules list instead, but that didn't seem right either.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to edorum2|
|-|-|

Hi,

I am confused.

What is the purpose/scope of this ticket/patch ?

The patch only contains a newsfile and I don't see the changes m.entioned in your last comment.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|edorum2 removed owner|
|-|-|

I messed up the first patch, it didn't have all the changes and just had the topfile. 
The second one has everything.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [45289]) Branching to py3-t.web.guard-7974.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to edorum2|
|-|-|

The changes are good.

The tests are failing on Fedora 19 / py3.3 -  https://buildbot.twistedmatrix.com/builders/fedora19-py3.3/builds/3301/steps/shell/logs/stdio

I have triggered a new tests run ... see latest results here https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/py3-t.web.guard-7974


The tests are still failing on py3.3 ... but are green on all other systems.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|edorum2 removed owner|
|-|-|

The base64 codec does not exists on Python 3.3, but it does on 3.4 and 2.7.
Switched to using the base64 module which works on all three. 
New patch should pass on Fedora as well.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

(In [45299]) Branching to py3-t.web.guard-7974-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to edorum2|
|-|-|

> The base64 codec does not exists on Python 3.3, but it does on 3.4 and 2.7.

What a mess :(


-------

For twisted/web/guard.py I think that we can take the opportunity to define where its tests are located.

I have already updated the header.

------

In dist3.py I think that we need to list all individual modules. Listing a package will not automatically include all the modules inside the package.

Right now we don't have py3 tests running inside the virtualenv with an installed twisted ... but we will have this soon and then the test will start to fail.

I have already made those changes.

--------

Waiting for test to end and then will merge.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set status to `closed`|
|-|-|

(In [45305]) Merge py3-t.web.guard-7974-2: Port twisted.web.guard to Python 3.

Author: edorum2
Reviewer: adiroiban
Fixes: [#7974](https://github.com/twisted/twisted/issues/7974)
